File: 1026_max_diff_between_node_and_ancestor.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxAncestorDiff(root: Optional[TreeNode]) -> int:
        def dfs(node: TreeNode, min_val, max_val, max_diff):
            res = max([max_diff, abs(min_val-node.val), abs(max_val-node.val)])
            if node.left:
                res = dfs(node.left, min(min_val, node.val), max(max_val, node.val), res)
            if node.right:
                res = dfs(node.right, min(min_val, node.val), max(max_val, node.val), res)
            return res
        
        res = 0
        if root.left:
            res = dfs(root.left, root.val, root.val, res)
        if root.right:
            res = dfs(root.right, root.val, root.val, res)
        return res

File: 1026_max_diff_between_node_and_ancestor.cpp
#include <algorithm>

using namespace std;
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        int max_diff = 0;
        dfs(max_diff, root->left, root->val, root->val);
        dfs(max_diff, root->right, root->val, root->val);
        return max_diff;
    }

    void dfs(int& max_diff, TreeNode* node, int max_val, int min_val) {
        if (node == nullptr) {
            max_diff = max(max_diff, max_val-min_val);
            return;
        }
        if (node->val < min_val) {
            min_val = node->val;
        } else if (node->val > max_val) {
            max_val = node->val;
        } 
        dfs(max_diff, node->left, max_val, min_val);
        dfs(max_diff, node->right, max_val, min_val);
    }
};